source-oracle/backfill.go

```diff
@@ -183,6 +208,122 @@ func (db *oracleDatabase) WatermarksTable() sqlcapture.StreamID {
 	return sqlcapture.JoinStreamID(bits[0], bits[1])
 }
 
+func (db *oracleDatabase) fetchBackfillRowIDRange(ctx context.Context, schemaName string, tableName string) error {
+	var streamID = sqlcapture.JoinStreamID(schemaName, tableName)
+	if _, ok := db.backfillRowIDRanges[streamID]; ok {
+		return nil
+	}
+
+	// This query starts by reading dba_extents which has information on data blocks of tables
+	// and splits those blocks into 1g buckets by calculating how many blocks are required to fill
+	// 1gb based on db_block_size, then dividing up the blocks into buckets using width_bucket,
+	// calculating the range begin end values necessary to construct rowids and
+	// finally generating ROWID ranges based on these buckets so we can do backfills 1gb at a time
+	var query = fmt.Sprintf(`
+    with extents_data as (
+      select o.data_object_id, e.file_id, e.block_id, e.blocks
+      from dba_extents e
+      join all_objects o
+      on (e.owner, e.segment_name, e.segment_type) = ((o.owner, o.object_name, o.object_type))
+        and decode(e.partition_name, o.subobject_name, 0, 1) = 0
+      where e.segment_type in ('TABLE', 'TABLE PARTITION', 'TABLE SUBPARTITION')
+        and e.owner = '%s'
+        and e.segment_name = '%s'
+    )
+    , extents_with_sums as (
+      select sum(blocks) over() total_blocks,
+        sum(blocks) over(order by data_object_id, file_id, block_id) - blocks cumul_prev_blocks,
+        sum(blocks) over(order by data_object_id, file_id, block_id) cumul_current_blocks,
+        greatest(1, FLOOR((sum(blocks) over())/(SELECT (1024*1024*1024) / TO_NUMBER(value) FROM v$parameter WHERE name='db_block_size'))) num_buckets,
```

Comment:
addition here from me to calculate number of buckets based on how many 1GB buckets would be required to cover data of the table